### DISCUSSION TITLE
Looping Constructs

### DISCUSSION THREAD

**Discussion by eliaslfox** [coalton-lang/coalton#847] [Ideas and Proposals]
[Moved from #840]

I had some initial thoughts about adding looping constructs to Coalton. This isn't a concrete plan for implementation.

Possible looping constructs could include:
* loop/forever
* while
* until (the inverse of while)
* c style for loops
* rust style iterator based for loops

I think whatever looping constructs get added, they should support break and continue.

A few other looping related feature that are worth considering:

[Breaking from `loop` with a value](https://doc.rust-lang.org/rust-by-example/flow_control/loop/return.html) like in Rust.

```rust
let mut counter = 0;

let result = loop {
    counter += 1;

    if counter == 10 {
        break counter * 2;
    }
};
```

[Breaking and continue with named blocks](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/label#using_a_labeled_break_with_for_loops) like in JavaScript and many other languages.

```javascript
let i, j;

// The first for statement is labeled "loop1"
loop1: for (i = 0; i < 3; i++) {
  // The second for statement is labeled "loop2"
  loop2: for (j = 0; j < 3; j++) {
    if (i === 1 && j === 1) {
      break loop1;
    }
    console.log(`i = ${i}, j = ${j}`);
  }
}

// Logs:
// i = 0, j = 0
// i = 0, j = 1
// i = 0, j = 2
// i = 1, j = 0
```

I also want to design something with a plan for how it can be optimized. Ideally the following two examples would compile identically.

```lisp
;; Coalton
(for (x (iter:into-iter (make-list 1 2 3 4 5)))
  (f x))

;; Lisp
(dolist (x (list 1 2 3 4 5))
  (f x))
```

**Comment by simendsjo** [comment]
F# has Computational Expressions where you can define how looping works for
instance. Programmable let, for, etc is quite powerful. See https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/computation-expressions#creating-a-new-type-of-computation-expression
